omegaconf/omegaconf.py

```diff
@@ -386,13 +388,48 @@ def register_resolver(
               of its inputs), and triggers a warning
             - `False` is the new behavior (the resolver can take non-string inputs), and
               will become the default in the future
+
+        If provided, `config_arg` should be the name of a keyword (typically keyword-only)
+        argument of `resolver` of type `BaseContainer`, that will be bound to the config
+        root when the resolver is called. This allows performing arbitrary operations on
+        the config from within the resolver. See `env()` for an example.
+
+        Similarly, `parent_arg` can be used to bind the corresponding keyword argument
+        of `resolver` (of type `Optional[Container]`) to the parent of the key being
+        processed when the resolver is called. This can be useful for operations involving
+        other config options relative to the current key.
+
+        `use_cache` indicates whether the resolver's outputs should be cached. When not
+            provided, it defaults to `True` unless either `config_arg` or `parent_arg` is
+            used. In such situations it defaults to `False` and the user is warned to
+            explicitly set `use_cache=False` to make it clear that no caching is done
+            (currently caching is not supported when using `config_arg` or `parent_arg`).
```

Comment:
The dependencies between the arguments is making this function hard to understand, use, and reason about.

I am starting to think that maybe keeping this function with the current signature and behavior, and adding a second one that introduces all the new goodies is a better path.

```
def register_resolver_ex(
        name: str,
        resolver: Resolver,
        use_cache: Optional[bool] = None,):
```

1. args_as_strings is not supported and is always false.
2. config_arg is not supported an is always passed (a part of the required resolver signature).
3. parent_arg is not supported (the user can access it via cfg._get_root()
it's currently private but we can make it a formal API:
```
OmegtaConf:
  def get_root(cfg) -> Optional[Container]:
    return cfg._get_root() # with some extra tests and stuff
```

This will allow us to keep `register_resolver` simple (no cache control even), while extending the API to support the more sophisticated behavior.
furthermore, we can deprecate `register_resolver` and point users to a migration guide page (or issue).
Once everyone are migrated to the new register_resolver API, we can rename it back to register_resolver in a future deprecation.

This is a big change though. What do you think?
Happy to have a VC with you to discuss.

Reply:
I wouldn't mind changing it but I have a couple of concerns:

* For (3) -- parent_arg is not supported -- I don't think `cfg._get_root()` is a 1-to-1 replacement, because the parent may be deep inside the config ==> finding the parent would require knowing the path from the root, which would make relative access difficult (ex: a `get_sibling` resolver as I added in the [docs](https://github.com/odelalleau/omegaconf/blob/grammar_interpolations/docs/source/usage.rst))

* If we allow people to set `use_cache=True` while giving them access to the root config, this could lead to surprising behavior (because the resolver could return the cached value, even though it shouldn't due to the config having been updated).

I don't feel too strongly about these though, so if you prefer it that way I don't mind changing it.

Reply:
Gave it a bit more thought:

* For (3) I just realized that probably you meant keeping access to the parent and not the whole config? (and one could get access to the whole config with `OmegaConf.get_root(parent)`)

* I just noticed that you used `use_cache=None` in the signature of `register_resolver_ex`, I guess this is for deprecation purpose (raising a warning if `use_cache` is not specified). This makes me wonder if it might be better to actually keep the name `register_resolver()`, but use the old version if `use_cache` is not provided, and use the new version when `use_cache` is set. The main advantages would be (i) cleaner code, and (ii) no need to go through another renaming pass later. The main disadvantage that I can see is that it is less explicit.

* Forcing people to set `use_cache` actually alleviates my second concern above (surprising caching behavior when accessing the config), since users would need to explicitly enable the cache.

* If we keep the idea of a new function, what about `register_resolver_new()` instead of `register_resolver_ex()`? (just to make it clearer what it is -- I'm not 100% sure what `_ex` stands for, extended maybe?)

Reply:
1. meant always passing the node of the interpolation (or it's parent if it makes more sense, not sure).
We can access the parent node, or the root node from any node.
2,3. `use_cache = None` is a copy paste error. It should probably just default to True. I am not sure it covers the full spectrum of the problems though:
The person registering the config resolver may not be the person editing the config and making it depend on things we can't easily cache (other nodes, other resolvers even?)

4. ex is extended. that suffix is commonly used to extend Windows APIs by adding a new function that will stay forever.
In this case we can eventually deprecate the simpler one and rename. I don't like _new in APIs because it feels like what you would have while you develop and not a real API.

Reply:
1. Ok. In my current implementation I'm using the parent because it's already available in the code, while the node itself is not. So it was really an "ease of implementation" choice and it can be revisited.

2. Not sure exactly what you mean here. The person registering the resolver would typically be the person who wrote it (or at least they should know what it's doing), so they should know if it's doing things that are not cache-friendly. My concern with `use_cache=True` being the default is that people may not pay attention to it, and write resolvers that should not use the cache but forget to register them with `use_cache=False`.

4. Ok

Bottom line: I'm happy to go with your suggestion if you still feel like it's the best direction to go. My concern raised in point 2 remains but it's not a major one. The deprecation plan would thus be as follows:

* In 2.1, warn people that they should be using `register_resolver_ex()`
* In 2.2, have `register_resolver()` raise a clear exception
* In 2.3, remove `register_resolver()`
* In 2.4, re-introduce `register_resolver()` as an alias to `register_resolver_ex()`, and warn people that they should use it from now on
* In 2.5, have `register_resolver_ex()` raise a clear exception
* In 2.6, remove `register_resoler_ex()`

That's a lot of steps but they are all easy to go through -- just need to make sure we don't forget about it...

Would that work for you?

Reply:
Answering again because GitHub sucks.
1. I think passing the parent of the node is okay.
2. The person writing the config is often not the person writing the resolver. example : env resolver and Hydra's now resolver.
I am not sure what we are talking about here anyway now. we can discuss the default caching behavior on the new function without less legacy baggage to confuse us.
4.
Shorter migration plan, let me know if you have concerns:
```
2.1:
- add legacy_register_resolver() which is the current register_resolver().
- introduce new_register_resolver()
- deprecated register_resolver() telling people to use new_register_resolver() or legacy_register_resolver() if they want a quick temporary fix.

2.2
No one is now using register_resolver()
- Move logic in new_register_resolver  to register_resolver and deprecate new_register_resolver
- Deprecated legacy_register_resolver() (this function will be removed in OmegaConf 2.3, use register_resolver)

2.3:
- delete both legacy_register_resolver() and new_register_resolver.
```

Reply:
Done in 4f0db2d

Regarding the default value of `use_cache`, let me know if you would like to change it (for now I kept the cache enabled by default)